### DISCUSSION TITLE
Unit tests for progress bar in pytest

### DISCUSSION THREAD

**Discussion by EdLeafe** [Textualize/rich#1093] [Q&A] [answered]
I have been working on adding `rich` to my CLI application by wrapping calls to rich in my output classes, and am writing unit tests to verify that it is working correctly. I use `pytest` as my testing framework, and use the `capsys` fixture for capturing and evaluating output. So far it's been great, but now I'm trying to unit test the progress bar code, and nothing is showing up in the captured output. I have verified that if I remove `capsys`, the progress bar is correctly written to the screen, but that is useless for testing.

To rule out that it was something in my wrappers that was the issue, I created some very simple unit tests: one with `capsys`, one with `capsysbinary`, and one with no capture just to demonstrate that it does actually write to the terminal output.

Can anyone help explain what I need to do to properly capture the output from the progress bar? Oh, I also tried creating a `Console` with `record=True`, and then checking the value from `export_text()`, but it's always empty, too.

```
import time

import pytest
from rich.console import Console
from rich.progress import BarColumn, Progress

TOTAL = 30
STEP = 5
LABEL = "Processing..."


def test_progress_bar(capsys):
    console = Console()
    progress = Progress(
        "[progress.description]{task.description}",
        BarColumn(),
        "[magenta]{task.completed} of {task.total} completed",
        console=console,
    )
    curr = 0
    with progress:
        task = progress.add_task(LABEL, total=TOTAL)
        while curr < TOTAL:
            curr += STEP
            time.sleep(0.1)
            progress.update(task, advance=STEP)
    cap = capsys.readouterr()
    assert LABEL in cap.out


def test_progress_bar_binary(capsysbinary):
    console = Console()
    progress = Progress(
        "[progress.description]{task.description}",
        BarColumn(),
        "[magenta]{task.completed} of {task.total} completed",
        console=console,
    )
    curr = 0
    with progress:
        task = progress.add_task(LABEL, total=TOTAL)
        while curr < TOTAL:
            curr += STEP
            time.sleep(0.1)
            progress.update(task, advance=STEP)
    cap = capsysbinary.readouterr()
    assert LABEL in cap.out.decode()


def test_progress_bar_no_capsys():
    console = Console()
    progress = Progress(
        "[progress.description]{task.description}",
        BarColumn(),
        "[magenta]{task.completed} of {task.total} completed",
        console=console,
    )
    curr = 0
    with progress:
        task = progress.add_task(LABEL, total=TOTAL)
        while curr < TOTAL:
            curr += STEP
            time.sleep(0.1)
            progress.update(task, advance=STEP)
    # Can't assert anything, since we haven't captured the output.
    # But you can see it in the terminal when you run the test.
    assert True
```

**Comment by willmcgugan** [comment, marked as answer]
Hi Ed,

I think I know what is going on. Rich will hide things like progress bars if it detects it is not writing to a terminal. This is because you don't want all the frames of progress bars written out if you are piping to a file for example.

You can add `force_terminal=True` to the Console constructor to bypass that and render the progress bars.

**Reply by EdLeafe** [reply]
That did it! Thanks for the quick response!

**Reply by MaKaNu** [reply]
EDIT: SOLVED. I am stupid. It works like a charm if I build my test actually the way I have planned my and used live.
For context. I have missed the live context manager for `Live` and I did a thing in the way I construct Progresses.
If someone is interested in the "THING" take a look here:  https://github.com/MaKaNu/PytorchImagePipeline/blob/experiment_sam2fcn/notebooks/experiments_core_rich.py#L351

I have a similar issue, but I use `Live(Group(*list_of_progresses))` as context manager. I have tried already to construct the progresses and the `Live` with a `Console(force_terminal=True)`, but no success.

I have tested the first example from @EdLeafe and actually forgot the `force_terminal=True` in the first place, and it worked like expected.

EDIT: I tried again with `force_terminal=True` and it worked, but this time I got the all the process instead just the final result.
